src/tree.rs

```diff
@@ -1423,6 +1423,81 @@ impl<T> Tree<T> {
     }
 }
 
+impl<T> PartialEq for Tree<T>
+where
+    T: PartialEq,
+{
+    fn eq(&self, other: &Tree<T>) -> bool {
+        if self.nodes.iter().filter(|x| x.is_some()).count()
+            != other.nodes.iter().filter(|x| x.is_some()).count()
+        {
+            return false;
+        }
+
+        for (node1, node2) in self
+            .nodes
+            .iter()
+            .filter_map(|x| (*x).as_ref())
+            .zip(other.nodes.iter().filter_map(|x| (*x).as_ref()))
+        {
+            let parent1_node = node1.parent.as_ref().and_then(|x| self.get(x).ok());
+            let parent2_node = node2.parent.as_ref().and_then(|x| other.get(x).ok());
```

Comment:
Question: do you know why we have a vector of Option<Node> for nodes in Tree instead of just having a vector of Node? I'm questioning all the logic about the free_ids too here.